tools/terraform-bundle/package.go

```diff
@@ -221,36 +154,41 @@ func (c *PackageCommand) Run(args []string) int {
 		}
 	}()
 
-	for _, file := range files {
-		if file.IsDir() {
-			// should never happen unless something tampers with our tmpdir
-			continue
-		}
+	// recursively walk the workDir to get a list of all binary filepaths
+	err = filepath.Walk(workDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
+			}
+			fn := info.Name()
 
-		fn := filepath.Join(workDir, file.Name())
-		r, err := os.Open(fn)
-		if err != nil {
-			c.ui.Error(fmt.Sprintf("Failed to open %s: %s", fn, err))
-			return 1
-		}
-		hdr, err := zip.FileInfoHeader(file)
-		if err != nil {
-			c.ui.Error(fmt.Sprintf("Failed to add zip entry for %s: %s", fn, err))
-			return 1
-		}
-		hdr.Method = zip.Deflate // be sure to compress files
-		w, err := outZ.CreateHeader(hdr)
-		if err != nil {
-			c.ui.Error(fmt.Sprintf("Failed to add zip entry for %s: %s", fn, err))
-			return 1
-		}
-		_, err = io.Copy(w, r)
-		if err != nil {
-			c.ui.Error(fmt.Sprintf("Failed to write %s to bundle: %s", fn, err))
-			return 1
-		}
-	}
+			// provider plugins need to be created in the same relative directory structure
+			absPath, _ := filepath.Abs(path)
+			relPath, _ := filepath.Rel(workDir, absPath)
+			relZip, err := outZ.Create(relPath)
```

Comment:
I can't find anything that sets the mode of the file in the zip archive, i.e. the equivalent of this code: https://github.com/hashicorp/terraform/blob/8d713375963fba27b4e69795ec2fb736d248d993/tools/terraform-bundle/package.go#L236-L242

I imagine this would result in a missing `+x` bit in the bundled binaries, but I haven't tested locally. Is this incorrect?

Reply:
Good question! I had no _idea_ that's what that code was doing, to be completely honest. I will look into that, thank you so much!

Reply:
fixed thank you again!